File: 2022/day17.cpp
// https://adventofcode.com/2022/day/17

#include "../common/common.h"
#include "../common/rust.h"

#include <algorithm>
#include <array>
#include <functional>
#include <map>
#include <print>
#include <ranges>

namespace stdr = std::ranges;
namespace stdv = std::views;

using rock_index_t = i64;
using Input = Vec<int>;
using grid_point = aoc::point_type<rock_index_t>;
constexpr let rock_size = 5;
using rock_t = std::array<grid_point, rock_size>;

fn parse(String const& filename) -> Input {
  auto lines = aoc::views::read_lines(filename);
  return str{*stdr::begin(lines)} |
         stdv::transform([](char c) { return (c == '<') ? -1 : 1; }) |
         aoc::ranges::to<Input>();
}

fn move_rock(rock_t& rock, grid_point delta) {
  for (grid_point& p : rock) {
    p += delta;
  }
}

template <rock_index_t num_rocks>
fn solve_case(Input const& pattern) -> rock_index_t {
  constexpr let use_sparse_grid = num_rocks > 10000;

  constexpr int left_most_index = 0;
  constexpr int right_most_index = 4;
  constexpr int top_most_index = 1;
  constexpr int bottom_most_index = 3;

  constexpr let initial_rock_displacement = grid_point{2, 3};

  constexpr let rock_structures = std::invoke([&] {
    std::array<rock_t, 5> rocks;
    rocks[0] = {grid_point{0, 0}, grid_point{1, 0}, grid_point{2, 0},
                grid_point{3, 0}, grid_point{3, 0}};
    rocks[1] = {grid_point{0, 1}, grid_point{1, 2}, grid_point{1, 1},
                grid_point{1, 0}, grid_point{2, 1}};
    rocks[2] = {grid_point{0, 0}, grid_point{2, 2}, grid_point{2, 0},
                grid_point{1, 0}, grid_point{2, 1}};
    rocks[3] = {grid_point{0, 1}, grid_point{0, 3}, grid_point{0, 0},
                grid_point{0, 0}, grid_point{0, 2}};
    rocks[4] = {grid_point{0, 0}, grid_point{0, 1}, grid_point{0, 0},
                grid_point{1, 0}, grid_point{1, 1}};

    // Each rock appears so that its left edge is two units away
    // from the left wall and its bottom edge is three units above
    // the highest rock in the room (or the floor, if there isn't one).
    for (auto& rock : rocks) {
      move_rock(rock, initial_rock_displacement);
    }
    return rocks;
  });
  constexpr let num_rock_types = rock_structures.size();

  constexpr let empty_char = ' ';
  constexpr let wall_char = '|'; // floor and side walls
  constexpr let rock_char = '#'; // settled rocks
  constexpr let chamber_width = 7;

  // Just use rocks as the walls, no need to do bounds checking
  using chamber_row_t = std::array<char, chamber_width + 2>;

  auto chamber = std::invoke([] {
    if constexpr (use_sparse_grid) {
      return aoc::sparse_grid<char, char{}, grid_point, chamber_row_t>{};
    } else {
      return aoc::grid<char, chamber_row_t, std::vector<char>>{};
    }
  });

  // And use rocks for the floor
  {
    auto floor = chamber_row_t{};
    floor.fill(wall_char);
    chamber.add_row(floor);
  }

  constexpr let empty_new_row = std::invoke([&] {
    chamber_row_t new_row;
    new_row.fill(empty_char);
    new_row[0] = new_row[new_row.size() - 1] = wall_char;
    return new_row;
  });

  let try_move_rock = [&](rock_t& falling_rock, grid_point delta) -> bool {
    move_rock(falling_rock, delta);
    let intersects_with_rock =
        stdr::any_of(falling_rock, [&](const grid_point& rp) {
          return chamber.at(rp.y, rp.x) != empty_char;
        });
    if (intersects_with_rock) {
      // Invalid movement, move it back
      move_rock(falling_rock, -delta);
      return false;
    }
    return true;
  };

  auto current_height = rock_index_t{1};
  int pattern_index = 0;
  let rock_solver = [&](rock_index_t rock_index) {
    auto falling_rock =
        rock_structures[static_cast<usize>(rock_index % num_rock_types)];
    let current_rock_height =
        falling_rock[top_most_index].y - falling_rock[bottom_most_index].y + 1;

    // Do initial falling stage without a chamber
    for (let _ : Range{0, initial_rock_displacement.y}) {
      // Only do horizontal movement in  this loop
      auto sideways = pattern[pattern_index];
      auto leftmost = falling_rock[left_most_index].x + sideways;
      auto rightmost = falling_rock[right_most_index].x + sideways;
      if ((leftmost >= 0) && (rightmost < chamber_width)) {
        move_rock(falling_rock, grid_point{sideways, 0});
      }
      pattern_index = (pattern_index + 1) % pattern.size();
    }

    { // Place falling rock into the chamber
      let missing_rows = current_height +
                         current_rock_height -
                         static_cast<rock_index_t>(chamber.num_rows());
      for (rock_index_t r = 0; r < missing_rows; ++r) {
        chamber.add_row(empty_new_row);
      }
      // This move includes the vertical movement from the initial falling stage
      // It also includes the left chamber wall
      move_rock(falling_rock,
                grid_point{1, current_height - initial_rock_displacement.y});
    }

    // Move rock until it settles
    loop {
      auto sideways = pattern[pattern_index];
      try_move_rock(falling_rock, grid_point{sideways, 0});
      if (!try_move_rock(falling_rock, grid_point{0, -1})) {
        // Rock settled, engrave it into the chamber
        for (const grid_point& rp : falling_rock) {
          chamber.modify(rock_char, rp.y, rp.x);
        }
        current_height =
            std::max(current_height, falling_rock[top_most_index].y + 1);
        pattern_index = (pattern_index + 1) % pattern.size();
        break;
      }
      pattern_index = (pattern_index + 1) % pattern.size();
    }
  };

  // Cycle detection: the simulation is periodic once the same
  // (rock_type, pattern_index, surface_hash) state recurse.
  //
  // rock_type and pattern_index alone are not sufficient:
  // two states can share those while the chamber surface is shaped differently
  // (rocks settled differently in earlier iterations).
  // Including a hash of the top 30 rows makes false positives
  // practically impossible, since 30 rows is well beyond
  // how deep any rock can reach when settling.
  //
  // Once a repeated state is found:
  //   cycle_rocks: how many rocks fit in one period
  //   cycle_height: how much the tower grows per period
  // We skip as many full periods as possible and simulate the remainder

  // State key: what uniquely identifies the simulation state
  // rock_type, pattern_index, surface_hash
  using CycleKey = std::tuple<int, int, size_t>;
  // State value: where this state was last seen
  struct SeenState {
    rock_index_t rock_index;
    rock_index_t height;
  };

  // Polynomial rolling hash of the top 30 chamber rows.
  // Each character is folded in as: hash = hash * 131 + c.
  // 131 is a prime larger than the printable ASCII range,
  // which spreads bits well and keeps collisions rare.
  // A false positive (two different surfaces producing the same hash)
  // would yield a wrong answer, but with a 64-bit result
  // and only thousands of states visited before a cycle is found,
  // the collision probability is negligible.
  let surface_hash = [&] {
    size_t hash = 0;
    for (rock_index_t k = 0; (k < 30) && ((current_height - 1 - k) >= 0); ++k) {
      for (char c : chamber.get_row(current_height - 1 - k)) {
        hash = hash * 131 + static_cast<size_t>(c);
      }
    }
    return hash;
  };

  auto seen = std::map<CycleKey, SeenState>{};
  auto skipped_height = rock_index_t{};
  bool cycle_found = false;

  for (rock_index_t index = 0; index < num_rocks; ++index) {
    // Only compute the key and update the map until the cycle is detected,
    // afterwards just simulate the remaining leftover rocks.
    if (!cycle_found) {
      let rock_type = static_cast<int>(index % num_rock_types);
      let key = CycleKey{rock_type, pattern_index, surface_hash()};

      if (seen.contains(key)) {
        let[prev_i, prev_height] = seen[key];
        let cycle_rocks = index - prev_i;
        let cycle_height = (current_height - 1) - prev_height;
        let full_cycles = (num_rocks - index) / cycle_rocks;
        skipped_height = full_cycles * cycle_height;
        index += full_cycles * cycle_rocks;
        cycle_found = true;
        if (index >= num_rocks) {
          break;
        }
      } else {
        seen[key] = SeenState{
            .rock_index = index,
            .height = current_height - 1,
        };
      }
    }

    rock_solver(index);
  }

  // Physical height of simulated rocks
  // plus the height of all skipped periods
  return current_height - 1 + skipped_height;
}

int main() {
  std::println("Part 1");
  let example = parse("day17.example");
  AOC_EXPECT_RESULT(17, solve_case<10>(example));
  AOC_EXPECT_RESULT(3068, solve_case<2022>(example));
  let input = parse("day17.input");
  AOC_EXPECT_RESULT(3085, solve_case<2022>(input));

  std::println("Part 2");
  AOC_EXPECT_RESULT(1514285714288, solve_case<1000000000000>(example));
  AOC_EXPECT_RESULT(1535483870924, solve_case<1000000000000>(input));

  AOC_RETURN_CHECK_RESULT();
}

File: rust/2022/src/day17.rs
use std::collections::HashMap;

type Input = Vec<i32>;

fn parse(filename: &str) -> Input {
    aoc::file::read_lines(filename)
        .into_iter()
        .next()
        .unwrap()
        .chars()
        .map(|c| if c == '<' { -1 } else { 1 })
        .collect()
}

// Each rock is 5 (x, y) points, some possibly duplicated, arranged so that
// specific indices are the boundary points used for fast collision checks.
const ROCK_SIZE: usize = 5;
type Rock = [(i64, i64); ROCK_SIZE];
const LEFT: usize = 0; // index of leftmost point
const TOP: usize = 1; // index of topmost point
const BOTTOM: usize = 3; // index of bottommost point
const RIGHT: usize = 4; // index of rightmost point

const WIDTH: i64 = 7;
const ROW_SIZE: usize = (WIDTH as usize) + 2; // 7 wide + 2 side walls
type ChamberRow = [char; ROW_SIZE];

const EMPTY: char = ' ';
const WALL: char = '|'; // floor and side walls
const ROCK: char = '#'; // settled rocks

// Polynomial rolling hash of the top 30 chamber rows.
// Each character is folded in as: hash = hash * 131 + c.
// 131 is prime and larger than the printable ASCII range,
// which spreads bits well and keeps collisions rare.
// A false positive would yield a wrong answer,
// but with a 64-bit result and only thousands of states
// visited before a cycle is found,
// the collision probability is negligible.
fn surface_hash(chamber: &[ChamberRow], current_height: i64) -> u64 {
    let mut hash: u64 = 0;
    for k in 0i64..30 {
        let row = current_height - 1 - k;
        if row < 0 {
            break;
        }
        for &c in &chamber[row as usize] {
            hash = hash.wrapping_mul(131).wrapping_add(c as u64);
        }
    }
    hash
}

fn solve_case(pattern: &Input, num_rocks: i64) -> i64 {
    // Each rock's 5 points are pre-displaced by (2, 3) so that the left edge
    // starts 2 units from the left wall and the bottom edge starts 3 units
    // above the current pile height (as required by the puzzle rules).
    const INIT_DX: i64 = 2;
    const INIT_DY: i64 = 3;
    const NUM_ROCK_TYPES: i64 = 5;
    let rock_templates: [Rock; NUM_ROCK_TYPES as usize] = {
        let offsets = [
            [(0, 0), (1, 0), (2, 0), (3, 0), (3, 0)], // ####
            [(0, 1), (1, 2), (1, 1), (1, 0), (2, 1)], // plus
            [(0, 0), (2, 2), (2, 0), (1, 0), (2, 1)], // reverse-L
            [(0, 1), (0, 3), (0, 0), (0, 0), (0, 2)], // |
            [(0, 0), (0, 1), (0, 0), (1, 0), (1, 1)], // square
        ];
        offsets.map(|rock| rock.map(|(x, y)| (x + INIT_DX, y + INIT_DY)))
    };

    // Chamber rows indexed by y-coordinate: row 0 is the floor.
    let mut chamber: Vec<ChamberRow> = vec![[WALL; ROW_SIZE]];
    let empty_row: ChamberRow = {
        let mut r = [EMPTY; ROW_SIZE];
        r[0] = WALL;
        r[8] = WALL;
        r
    };

    let mut current_height: i64 = 1;
    let mut pattern_index: usize = 0;

    // Cycle detection: key = (rock_type, pattern_index, surface_hash).
    // When the same key recurs the full simulation state is periodic.
    let mut seen: HashMap<(usize, usize, u64), (i64, i64)> = HashMap::new();
    let mut skipped_height: i64 = 0;
    let mut cycle_found = false;

    let mut index: i64 = 0;
    while index < num_rocks {
        // Only compute the key and update the map until the cycle is detected;
        // afterwards just simulate the remaining leftover rocks.
        if !cycle_found {
            let rock_type = (index % NUM_ROCK_TYPES) as usize;
            let key = (
                rock_type,
                pattern_index,
                surface_hash(&chamber, current_height),
            );
            if let Some(&(prev_index, prev_height)) = seen.get(&key) {
                let cycle_rocks = index - prev_index;
                let cycle_height = (current_height - 1) - prev_height;
                let full_cycles = (num_rocks - index) / cycle_rocks;
                skipped_height = full_cycles * cycle_height;
                index += full_cycles * cycle_rocks;
                cycle_found = true;
                if index >= num_rocks {
                    break;
                }
            } else {
                seen.insert(key, (index, current_height - 1));
            }
        }

        let mut rock = rock_templates[(index % NUM_ROCK_TYPES) as usize];
        let rock_height = rock[TOP].1 - rock[BOTTOM].1 + 1;

        // Initial falling stage: consume one jet per row
        // for the 3 empty rows above the pile, without collision checking
        // (the rock is guaranteed to be above everything at this point).
        for _ in 0..INIT_DY {
            let sideways = pattern[pattern_index] as i64;
            let leftmost = rock[LEFT].0 + sideways;
            let rightmost = rock[RIGHT].0 + sideways;
            if leftmost >= 0 && rightmost < WIDTH {
                for p in rock.iter_mut() {
                    p.0 += sideways;
                }
            }
            pattern_index = (pattern_index + 1) % pattern.len();
        }

        // Extend the chamber to fit the rock,
        // then shift into chamber coordinates:
        // +1 for the left wall, y to current_height - INIT_DY
        let missing = current_height + rock_height - chamber.len() as i64;
        for _ in 0..missing.max(0) {
            chamber.push(empty_row);
        }
        for pos in rock.iter_mut() {
            pos.0 += 1;
            pos.1 += current_height - INIT_DY;
        }

        // Move rock until it settles
        loop {
            let sideways = pattern[pattern_index] as i64;
            // Try horizontal jet
            let can_move_h = rock
                .iter()
                .all(|&(x, y)| chamber[y as usize][(x + sideways) as usize] == EMPTY);
            if can_move_h {
                for pos in rock.iter_mut() {
                    pos.0 += sideways;
                }
            }
            pattern_index = (pattern_index + 1) % pattern.len();

            // Try to fall one step
            let can_fall = rock
                .iter()
                .all(|&(x, y)| chamber[(y - 1) as usize][x as usize] == EMPTY);
            if can_fall {
                for p in rock.iter_mut() {
                    p.1 -= 1;
                }
            } else {
                // Rock settled: engrave it and update height
                for &(x, y) in &rock {
                    chamber[y as usize][x as usize] = ROCK;
                }
                current_height = current_height.max(rock[TOP].1 + 1);
                break;
            }
        }

        index += 1;
    }

    // Physical height of simulated rocks plus the height of all skipped periods
    current_height - 1 + skipped_height
}

fn main() {
    println!("Part 1");
    let example = parse("day17.example");
    aoc::expect_result!(17, solve_case(&example, 10));
    aoc::expect_result!(3068, solve_case(&example, 2022));
    let input = parse("day17.input");
    aoc::expect_result!(3085, solve_case(&input, 2022));

    println!("Part 2");
    aoc::expect_result!(1514285714288_i64, solve_case(&example, 1_000_000_000_000));
    aoc::expect_result!(1535483870924_i64, solve_case(&input, 1_000_000_000_000));
}
